### ISSUE TITLE
Research how writing is performed

### ISSUE THREAD

**Issue by patstave** [patstave/Node-FlexitCS60-RS485#5]
We need a stable way to send data TO the control board. Is there a fixed time interval for sending? 
The electrical layer is only two wire (has to be half-duplex).

How do we send data without interfering with the data sent from the control board?

**Comment by danielcherubini** [comment]
Hey, sorry to bump this, but did you make any headway here?

**Comment by patstave** [comment]
Sorry no, the project is dead from my side

**Comment by MSkjel** [comment]
I havent actually gotten around to doing writing myself. What I know though is that the ventilation system controller itself seems to be requesting data from my CI600 in a master -> slave configuration. The ventilation system controller sends a request to the CI600 and the CI600 then responds with whatever it needs to.  I might be wrong here, but I just figured out how the packets themselves are validated. Does anybody have a CI66 and a RS485 adapter? Just a packet dump from a bus with a CI66 would help alot!

**Comment by patstave** [comment]
I agree that a packet dump from a system with CI66 would really be the way to go. Unfortunately i do not have access to it. 
It is a long time since i have looked at the packets. But i seem to remember that the master actually broadcast a lot of the information. Eg. no validation / transmission control. What i never fully figured out is how a set point change is actually transmitted from the slave to the master. There is so much "noise" and high transmission frequency on the bus. Really overkill baudrate and transmission rate for the application.

**Comment by MSkjel** [comment]
The thing is, I dont see any places the CI600 actually transmits any data without first receiveing any data. This is the only function that processes UART data for UART1. As you can see it only actually transmits data if it receives it first. I might be wrong here, as I havent really had the time to look into it more thoroughly. As you can see it configues the UART transceiver for sending, and then reconfigures it for receiveing afterwards. This is the only function this is done in. Flexit_HandleAndCreatePacket does alot of stuff and has a really messy control-flow. Some of the functions it calls are even loaded from flash, hence the pointers passed into it.

What is really weird though, and might confirm your hypothesis is that the setpoint for temperature isnt sent continously. Only when you actually change it on the CI600. What Im thinking though, is that the CS60 sends some kind of "Do you have any changes" request to the CI600, and then the CI600 somehow responds with yes, or the setpoint itself? I dont know, just throwing around ideas. I'll look into the code more this weekend!

```
void Flexit_Process_data(undefined2 length?,undefined2 reg_r1?)

{
  bool is_transmission_complete;
  bool bVar1;
  int shouldSend;
  int iVar2;
  undefined4 data_length;
  byte *current_byte;
  
  data_length = CONCAT22(length?,reg_r1?);
  shouldSend = Flexit_HandleAndCreatePacket?
                         (1,(ushort *)((int)&data_length + 2),UART1_Buffer,
                          (protocol_handler_function_pointers *)&PTR_func1+1_20000008);
  if (shouldSend == 1) {
    GPIO_Write(0x5,0x0);
    GPIO_Write(0x4,0x1);
    iVar2 = 0;
    do {
      current_byte = UART1_Buffer + iVar2;
      iVar2 = iVar2 + 1;
      USART_SendData((int)&Peripherals::USART1,(ushort)*current_byte);
      do {
        is_transmission_complete = FUN_0800eee0((ushort *)&Peripherals::USART1,0x80);
      } while (!is_transmission_complete);
    } while (iVar2 < (int)(data_length >> 0x10));
    do {
      bVar1 = FUN_0800eee0((ushort *)&Peripherals::USART1,0x40);
    } while (!bVar1);
    GPIO_Write(0x4,0x0);
    GPIO_Write(0x5,0x1);
  }
  return;
}
```

**Comment by MSkjel** [comment]
Still waiting on a MAX485 to test writing, but a WIP can be found on https://github.com/MSkjel/esphome-flexit-modbus-server

**Comment by MSkjel** [comment]
Ok, got writing working. For now just the mode and the setpoint temperature. This is just a normal modbus client -> server system. You need to implement your own modbus server to be able to write changes as the CS60 is just a modbus client. I dont think you can make this work with the Elfin, as it can only work as a modbus client. See my repository in the previous comment. It can probably be ported to anything else if anybody wants to. Now I just have to figure out what else we can do :)
